packages/peregrine/src/Toasts/useToastContext.js

```diff
@@ -0,0 +1,125 @@
+import React, { createContext, useContext, useReducer } from 'react';
+/**
+ * The state of the toast store.
+ * @typedef {Object} ToastState
+ * @property {Object} toasts - unordered map of id:toast
+ */
+const initialState = {
+    toasts: {}
+};
+
+const reducer = (prevState = initialState, action = {}) => {
+    const { type, payload } = action;
+
+    if (!type) {
+        throw new TypeError('action.type is required');
+    }
+
+    if (!payload.id) {
+        throw new TypeError('action.payload.id is required');
+    }
+
+    switch (type) {
+        case 'add': {
+            const { toasts } = prevState;
+            const { [payload.id]: prevToast } = toasts;
+
+            // If we are adding a toast that already exists we need to clear the
+            // old removal timeout effectively resetting the delete timer.
+            if (prevToast) {
+                window.clearTimeout(prevToast.removalTimeoutId);
+            }
+
+            const duplicate = !!prevToast;
+
+            // For duplicate toasts, do not update the timestamp to maintain
+            // order of toast emission.
+            const timestamp = duplicate ? prevToast.timestamp : Date.now();
+
+            // Use a random key to trigger a recreation of this component if it
+            // is a duplicate so that we can re-trigger the blink animation.
+            const key = duplicate ? Math.random() : payload.id;
+
+            const nextState = {
+                ...prevState,
+                toasts: {
+                    ...prevState.toasts,
+                    [payload.id]: {
+                        ...payload,
+                        timestamp,
+                        duplicate,
+                        key
+                    }
+                }
+            };
+
+            return nextState;
+        }
+        case 'remove': {
+            const nextState = {
+                ...prevState
+            };
+
+            // It is possible to attempt to delete a non-existant toast so let's
+            // gracefully handle that.
+            if (nextState.toasts[payload.id]) {
+                // Clear the old toast removal timeout incase an identical toast
+                // is added after this.
+                window.clearTimeout(
+                    nextState.toasts[payload.id].removalTimeoutId
+                );
+
+                // Delete the toast from the store.
+                delete nextState.toasts[payload.id];
+            }
+
+            // Do not blink on removal.
+            Object.keys(nextState.toasts).forEach(key => {
+                nextState.toasts[key].duplicate = false;
+            });
```

Comment:
On further review, I noticed this is actually mutating the toast objects, which we can't do. 😞 

```es6
const toasts = {
  a: { duplicate: true },
  b: { duplicate: true },
  c: { duplicate: false },
  d: { duplicate: false },
}

const mutate = (id) => {
  const nextToasts = { ...toasts }

  delete nextToasts[id]

  Object.keys(nextToasts).forEach(key => {
    nextToasts[key].duplicate = false
  })

  return nextToasts
}

test("shouldn't mutate objects", () => {
  const { a: before } = toasts
  const { a: after } = mutate("d")

  // sets `duplicate` to `false`, as intended
  expect(before.duplicate).toBe(true)   // ✅
  expect(after.duplicate).toBe(false)   // ✅

  // but doesn't change object identity
  expect(before).not.toBe(after)        // ❌
})
```

Reply:
This is something I did in an attempt to get the "blink" animation to work properly. Without that side effect, all the toasts were blinking because the toast container is essentially re-rendering/recomputing the list on every mutation. I'd really like to make this less complex.

Reply:
Edit: Actually wait, I think I remedied that by using the key. Why did I leave this in?